python/ql/lib/semmle/python/frameworks/internal/SubclassFinder.qll

```diff
@@ -0,0 +1,203 @@
+/**
+ * INTERNAL: Do not use.
+ *
+ * Has predicates to help find subclasses in library code. Should only be used to aid in
+ * the manual library modeling process,
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.ApiGraphs
+private import semmle.python.filters.Tests
+
+// very much inspired by the draft at https://github.com/github/codeql/pull/5632
+private module NotExposed {
+  // Instructions:
+  // This needs to be automated better, but for this prototype, here are some rough instructions:
+  // 0) get a database of the library you are about to model
+  // 1) fill out the `getAlreadyModeledClass` body below
+  // 2) quick-eval the `quickEvalMe` predicate below, and copy the output to your modeling predicate
+  class MySpec extends FindSubclassesSpec {
+    MySpec() { this = "MySpec" }
+
+    override API::Node getAlreadyModeledClass() {
+      // FILL ME OUT ! (but don't commit with any changes)
+      none()
+      // for example
+      // result = API::moduleImport("rest_framework").getMember("views").getMember("APIView")
+    }
+  }
+
+  predicate quickEvalMe(string newImport) {
+    newImport =
+      "// imports generated by python/frameworks/internal/SubclassFinder.qll\n" + "this = API::" +
+        concat(string newModelFullyQualified |
+          newModel(any(MySpec spec), newModelFullyQualified, _, _, _)
+        |
+          fullyQualifiedToAPIGraphPath(newModelFullyQualified), " or this = API::"
+        )
+  }
+
+  // ---------------------------------------------------------------------------
+  // Implementation below
+  // ---------------------------------------------------------------------------
+  //
+  // inherent problem with API graphs is that there doesn't need to exist a result for
+  // all the stuff we have already modeled... as an example, the following query has no
+  // results when evaluated against a django/django DB
+  //
+  // select API::moduleImport("django")
+  //       .getMember("contrib")
+  //       .getMember("admin")
+  //       .getMember("views")
+  //       .getMember("main")
+  //       .getMember("ChangeListSearchForm")
+  //
+  // therefore we use fully qualified names to capture new classes/new aliases.
+  //
+  // note that this implementation was originally created to help with automatically
+  // modeling packages in mind, and was just copied for this purpose. See
+  // https://github.com/github/codeql/pull/5632 for more discussion. I wanted to get
+  // this into the codeql-repo, so it could be of use when modeling 3rd party libraries,
+  // and save some manual effort.
```

Comment:
I would rewrite it slightly:
```suggestion
  // ---------------------------------------------------------------------------
  // Implementation below
  // ---------------------------------------------------------------------------
  // We are looking to find all subclassed of the already modelled classes, and
  // ideally we would identify an `API::Node` for each (then `toString` would give
  // the API path).
  //
  // However, an inherent problem with API graphs is that there doesn't need to exist a result for
  // all the stuff we have already modeled... as an example, the following query has no
  // results when evaluated against a django/django DB
  //
  // select API::moduleImport("django")
  //       .getMember("contrib")
  //       .getMember("admin")
  //       .getMember("views")
  //       .getMember("main")
  //       .getMember("ChangeListSearchForm")
  //
  // Therefore we use fully qualified names to capture new classes/new aliases, and transform these
  // into API paths, see `fullyQualifiedToAPIGraphPath`.
  //
  // Further automation is planned her, see
  // https://github.com/github/codeql/pull/5632 for more discussion.
```

Reply:
I didn't fully agree, but have rewritten based on your suggestion :+1: